cuda/dot_benchmark.cu

```diff
@@ -0,0 +1,197 @@
+#include <array>
+#include <cmath>
+#include <functional>
+#include <iomanip>
+#include <ios>
+#include <iostream>
+#include <random>
+#include <type_traits>
+
+
+#include "dot_kernels.cuh"
+#include "dot_memory.cuh"
+#include "memory.cuh"
+#include "utils.cuh"
+
+
+int main(int argc, char **argv)
+{
+    using ar_type = double;
+    using st_type = float;
+    using value_type = ar_type;
+
+    constexpr std::size_t max_size{535 * 1000 * 1000};
+    constexpr char DELIM{';'};
+
+    bool detailed_error{false};
+
+    const std::string use_error_string("--error");
+    if (argc == 2 && std::string(argv[1]) == use_error_string) {
+        detailed_error = true;
+    } else if (argc > 1) {
+        const std::string binary(argv[0]);
+        std::cerr << "Unsupported parameters!\n";
+        std::cerr << "Usage: " << binary << " [" << use_error_string << "]\n";
+        std::cerr << "With " << use_error_string
+                  << ":    compute detailed error of DOTs\n"
+                  << "Without parameters: benchmark different DOTs\n";
+        return 1;
+    }
+    std::default_random_engine rengine(42);
+    std::uniform_real_distribution<value_type> vector_dist(-1.0, 1.0);
+
+    auto ar_data = DotMemory<ar_type>(max_size, vector_dist, rengine);
+    auto st_data = DotMemory<st_type>(ar_data);
+
+    auto cublas_handle = cublas_get_handle();
+    cublas_set_device_ptr_mode(cublas_handle.get());
+
+    cudaDeviceProp device_prop;
+    CUDA_CALL(cudaGetDeviceProperties(&device_prop, 0));
+
+    auto ar_get_result = [&ar_data]() { return ar_data.get_result(); };
+    auto st_get_result = [&st_data]() {
+        return static_cast<ar_type>(st_data.get_result());
+    };
+
+    constexpr std::size_t benchmark_num{7};
+    constexpr std::size_t benchmark_reference{0};
+    using benchmark_info_t =
+        std::tuple<std::string, std::function<void(matrix_info, matrix_info)>,
+                   std::function<value_type()>>;
+    std::array<benchmark_info_t, benchmark_num> benchmark_info = {
```

Comment:
The main reason why I didn't is because I started using `std::array`, which can't infer the size.
I could definitively go for `std::vector` (or `std::map`) and replace all `std::array` with `std::vector`. I will check if I run into any problems, if not, I will use your approach.

Reply:
Changed everything to `std::vector` now instead of `std::array`.